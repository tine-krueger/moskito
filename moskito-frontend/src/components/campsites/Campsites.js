import getCampsites from '../../services/getCampsites';
import { useState, useEffect } from 'react';

export default function Campsites() {
    const [campsites, setCampsites] = useState([])

    useEffect(() => {
        getCampsites()
        .then(data => {
            setCampsites(data)
            console.log(data)
        })
        
        .catch(error => console.log(error))
    }, [])


    return (
        <div>
            {campsites.map(campsite => ( 
                <div key={campsite.id}>
                    <h2>{campsite.name}</h2>
                    <p>{campsite.street}</p>
                    <p>{campsite.postalCode} {campsite.place}</p>
                    <p>Tel: {campsite.telephone}</p>
                    <p>E-Mail: {campsite.email}</p>
                    {filterFeatures(campsite).map(feature => <li>{feature.type}</li>)}
                </div>
                )
            )}
        </div>
    )

    function filterFeatures(campsite) {
        const features =  campsite.features.filter(feature => feature.value)
        return features
    }
}
