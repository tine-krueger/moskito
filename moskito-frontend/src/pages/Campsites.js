import PropTypes from 'prop-types'
import CampsiteList from '../components/Campsitelist/CampsiteList'
import Header from '../components/Header/Header'
import Loading from '../components/Loading/Loading'
import Navigation from '../components/Navigation/Navigation'

Campsites.propTypes = {
    headline: PropTypes.string.isRequired,
    campsites: PropTypes.arrayOf(PropTypes.object).isRequired,
    errors: PropTypes.object,
    isLoading: PropTypes.bool.isRequired
}

export default function Campsites({headline, campsites, errors, isLoading}) {
    
    return (
        <>
            <Header children={headline}/>
            {isLoading ? <Loading/> : <CampsiteList isCampsite={true} campsites={campsites} errors={errors}/>}
            <Navigation/>
        </>
    )   
}
