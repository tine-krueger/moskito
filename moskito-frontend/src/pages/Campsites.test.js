import 'jest-styled-components'
import userEvent from '@testing-library/user-event'
import Campsites from './Campsites'
import renderWithRouter from '../testSetup/setupTests'

const mockHeadline = 'Headline'
const mockCampsites = [{
    id: 1,
    name: 'Test Campsite',
    pinned: false,
    features: [
        {
            id: 1,
            type: 'wlan',
            value: true
        },
        {
            id: 2,
            type: 'music',
            value: false
        }
    ]    
}]
const mockIsLoading = false
const mockDeleteTokens = jest.fn()
const mocksetAuthTokens = jest.fn()

const component = <Campsites
    headline={mockHeadline}
    campsites={mockCampsites}
    isLoading={mockIsLoading}
    />

jest.mock('../context/auth.js', () => ({
    ...jest.requireActual('../context/auth.js'),
    useAuth: () => ({
        deleteTokens: mockDeleteTokens,
        setAuthTokens: mocksetAuthTokens
    })
}))

describe('Campsites Page', () => {
    it('renders correctly', () => {
        const { container } = renderWithRouter(component)
        expect(container).toMatchSnapshot()
    })

    it('shows a headline, one campsite with like-button and the right true-features and a navigation', () => {
        const { getByRole, getByText, queryByText, getByTestId } = renderWithRouter(component)
        expect(getByRole('heading', {name: /Headline/i})).toBeInTheDocument()
        expect(getByRole('heading', {name: /Test Campsite/i})).toBeInTheDocument()
        expect(getByText(/kein W-Lan/i)).toBeInTheDocument()
        expect(queryByText(/keine Bar/i)).not.toBeInTheDocument()
        expect(getByTestId('like-button')).toBeInTheDocument()
        expect(getByRole('navigation')).toBeInTheDocument()

    })

    it('shows only the loading-spinner, when loading is true', () => {
        const { getByTestId, queryByRole } = renderWithRouter(
            <Campsites
            headline={mockHeadline}
            campsites={mockCampsites}
            isLoading={true}
            />
        )
        expect(getByTestId('loading-spinner')).toBeInTheDocument()
        expect(queryByRole('heading', {name: /Test Campsite/i})).not.toBeInTheDocument()
    })

    it('shows errors, when errors are existing', () => {
        const { queryByRole, getByText } = renderWithRouter(
            <Campsites
            headline={mockHeadline}
            campsites={mockCampsites}
            isLoading={mockIsLoading}
            errors={{errors: 'It has errors'}}
            />
        )
        expect(getByText(/It has Errors/i)).toBeInTheDocument()
        expect(queryByRole('heading', {name: /Test Campsite/i})).not.toBeInTheDocument()
    })

    it('navigates to bookmarks-page, when clicking on nav-button', () => {
        const { getByTestId, history } = renderWithRouter(component)
        userEvent.click(getByTestId('bookmarks'))
        expect(history.location.pathname).toBe('/bookmarks')
    })

    it('navigates to find-campsites-page, when clicking on nav-button', () => {
        const { getByTestId, history } = renderWithRouter(component)
        userEvent.click(getByTestId('find-campsite'))
        expect(history.location.pathname).toBe('/find-campsite')
    })

    it('logs out by deleting tokens, when clicking on nav-button', () => {
        const { getByTestId } = renderWithRouter(component)
        userEvent.click(getByTestId('logout'))
        expect(mocksetAuthTokens).toHaveBeenCalledTimes(1)
        expect(mockDeleteTokens).toHaveBeenCalledTimes(1)
    })
})