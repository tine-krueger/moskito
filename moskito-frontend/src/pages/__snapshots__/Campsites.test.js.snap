// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campsites Page renders correctly 1`] = `
.c6 {
  border-radius: 10px;
  display: inline-block;
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
  padding: .5em 1em;
  margin: .3em;
}

.c7 {
  position: absolute;
  right: 1em;
  top: 1em;
  height: 2.5em;
  width: 2.5em;
  border-radius: 10px;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 svg {
  width: 1.7em;
  height: 1.7em;
  stroke: var(--link);
  stroke-width: 3em;
  fill: var(--background);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-bottom: 3px;
}

.c2 {
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
  border-radius: 20px;
  margin: 2em 0;
  padding: .5em 1em;
  position: relative;
}

.c2 h2 {
  width: 75%;
  margin-left: .3em;
}

.c3 {
  margin-left: .5em;
}

.c4 p {
  position: relative;
  padding-left: 2em;
}

.c4 svg {
  width: 20px;
  height: 20px;
  color: var(--headline);
  position: absolute;
  top: 0;
  left: 0;
}

.c4 a {
  display: block;
  margin-top: 1em;
}

.c5 {
  list-style: none;
  padding-left: 0;
  position: relative;
  margin-top: 3em;
}

.c5::before {
  content: " ";
  display: inline-block;
  position: absolute;
  width: 90%;
  height: 5px;
  left: 5%;
  top: -25px;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  border-radius: 5px;
  box-shadow: 0.5px 0.5px 2px 0px rgba(0,0,0,0.4),-1.25px -1.25px 2px 0px rgba(255,255,255,0.5);
}

.c1 {
  margin-bottom: 6em;
  padding: 0 2em;
}

.c0 {
  width: 80%;
  margin: 2em auto 3em;
  padding: .25em .5em;
  border-radius: 20px;
  text-align: center;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
}

.c8 {
  margin: 0 auto;
  max-width: 500px;
  background-color: var(--background);
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-2px -2px 2px 0px rgba(255,255,255,0.5);
}

.c9 {
  padding-left: 0;
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c10 {
  text-align: center;
  border-radius: 10px;
  box-shadow: 2px 4px 8px -4px rgba(0,0,0,0.6),-1px -1px 2px 0px rgba(255,255,255,0.5);
  width: 2.3em;
  height: 2.3em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c10 button {
  background-color: transparent;
  border: none;
}

.c10 button:hover {
  outline: none;
}

.c10 a svg,
.c10 button svg {
  width: 1.5em;
  height: 1.5em;
  stroke: var(--link);
}

.c10 button svg {
  width: 1.7em;
  height: 1.7em;
}

.c10:hover,
.c10:active,
.c10:focus {
  box-shadow: inset 4px 4px 6px 0 rgba(0,0,0,0.2),inset -3px -4px 6px 0 rgba(255,255,255,0.3);
  outline: none;
}

<div>
  <div
    class="c0"
  >
    <h1>
      Headline
    </h1>
  </div>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <h2>
        Test Campsite
      </h2>
      <div
        class="c3"
      >
        <p>
          , 
           
        </p>
        <div
          class="c4"
        >
          <a
            rel="noopener noreferrer"
            target="_blank"
          />
        </div>
      </div>
      <ul
        class="c5"
      >
        <li
          class="c6"
        >
          kein W-Lan
        </li>
      </ul>
      <div
        class="c7"
        data-testid="like-button"
      >
        <svg
          data-testid="pin"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M878.3 392.1L631.9 145.7c-6.5-6.5-15-9.7-23.5-9.7s-17 3.2-23.5 9.7L423.8 306.9c-12.2-1.4-24.5-2-36.8-2-73.2 0-146.4 24.1-206.5 72.3-15.4 12.3-16.6 35.4-2.7 49.4l181.7 181.7-215.4 215.2a15.8 15.8 0 0 0-4.6 9.8l-3.4 37.2c-.9 9.4 6.6 17.4 15.9 17.4.5 0 1 0 1.5-.1l37.2-3.4c3.7-.3 7.2-2 9.8-4.6l215.4-215.4 181.7 181.7c6.5 6.5 15 9.7 23.5 9.7 9.7 0 19.3-4.2 25.9-12.4 56.3-70.3 79.7-158.3 70.2-243.4l161.1-161.1c12.9-12.8 12.9-33.8 0-46.8z"
          />
        </svg>
      </div>
    </div>
  </div>
  <nav
    class="c8"
  >
    <ul
      class="c9"
    >
      <li
        class="c10"
      >
        <a
          data-testid="find-campsite"
          href="/find-campsite"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
            />
          </svg>
        </a>
      </li>
      <li
        class="c10"
      >
        <a
          data-testid="bookmarks"
          href="/bookmarks"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M878.3 392.1L631.9 145.7c-6.5-6.5-15-9.7-23.5-9.7s-17 3.2-23.5 9.7L423.8 306.9c-12.2-1.4-24.5-2-36.8-2-73.2 0-146.4 24.1-206.5 72.3-15.4 12.3-16.6 35.4-2.7 49.4l181.7 181.7-215.4 215.2a15.8 15.8 0 0 0-4.6 9.8l-3.4 37.2c-.9 9.4 6.6 17.4 15.9 17.4.5 0 1 0 1.5-.1l37.2-3.4c3.7-.3 7.2-2 9.8-4.6l215.4-215.4 181.7 181.7c6.5 6.5 15 9.7 23.5 9.7 9.7 0 19.3-4.2 25.9-12.4 56.3-70.3 79.7-158.3 70.2-243.4l161.1-161.1c12.9-12.8 12.9-33.8 0-46.8z"
            />
          </svg>
        </a>
      </li>
      <li
        class="c10"
      >
        <button
          data-testid="logout"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
            />
            <polyline
              points="16 17 21 12 16 7"
            />
            <line
              x1="21"
              x2="9"
              y1="12"
              y2="12"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
